Open Source LLMs (#67)
pkeilbach/htwg-practical-nlp: docs/presentations/articles/OS_LLMs_Robert.md

Reviewer: [naming convention](https://pkeilbach.github.io/htwg-practical-nlp/presentations/presentations/#submission-via-pull-request) would be `open_source_llms.md`. No need to include your name ✌🏼

Author: Okay

> <span style="color:red; font-size: 40px; font-weight: bold;">Open-Source LLMs</span>

Reviewer: Please make it just a regular level 1 heading:

```suggestion
# Open-Source LLMs
```

> #### An Overview of Open Source LLMs: What Open Source Alternatives Exist to OpenAI's GPT-3.5 and How Can I Utilize Them?

Reviewer: no need for a subtitle. This will look strange in the outline:

![image](https://github.com/pkeilbach/htwg-practical-nlp/assets/28002199/9fea4d0c-6cf8-4809-aef8-82e4975d14ef)

```suggestion
This article gives an overview of Open Source LLMs: what Open Source alternatives exist to OpenAI's GPT-3.5 and how can we utilize them?
```

Author: thanks

Autor: Karoly Robert Hegyi

> Datum: 21.12.2023

Reviewer: The German version is not needed

Author: Karoly Robert Hegyi
Date: 21.12.2023

Open Source Large Language Models (OSLLMs) are powerful, open-source AI models for natural language processing tasks. Unlike proprietary models, they are transparent, facilitate collaboration, and foster innovation in applications such as translation, text generation, and chatbots.

## Availability of Large Language Models

They are distinguished into:

| Proprietary LLMs                         | Open Source LLMs                          |
| ---------------------------------------- | ----------------------------------------- |
| Owned by a company that controls its use | Free of charge and accessible to everyone |
| License fees required.                   | Free commercial use permitted.            |

Various licenses exist for open-source software, with the majority adhering to the Apache 2.0 standard.

!!! info Apache 2.0

> Allows users to use, distribute, modify, and distribute modified versions of the software for any purpose under the terms of the license, without worrying about license fees.

Reviewer: To use the [admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#changing-the-title)correctly, you need to watch out for the indentation and blank lines, and putting the title in quotes `"`

Also see [this](https://github.com/pkeilbach/htwg-practical-nlp/blob/6bd61b90f6a957234a0c3b4b132ed3bafb0ca63a/docs/lectures/minimum_edit_distance.md?plain=1#L363) example as a reference.

> !!! info Apache 2.0
> Allows users to use, distribute, modify, and distribute modified versions of the software for any purpose under the terms of the license, without worrying about license fees.

Reviewer: To use the [admonition](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#changing-the-title)correctly, you need to watch out for the indentation and blank lines, and putting the title in quotes `"`

```suggestion
!!! info "Apache 2.0"

    Allows users to use, distribute, modify, and distribute modified versions of the software for any purpose under the terms of the license, without worrying about license fees.
```

Also see [this](https://github.com/pkeilbach/htwg-practical-nlp/blob/6bd61b90f6a957234a0c3b4b132ed3bafb0ca63a/docs/lectures/minimum_edit_distance.md?plain=1#L363) example as a reference.

Reviewer: please adjust this throughout the document.

Author: Thanks, somehow every time i push the changes it automatically formats itself wrongly

## Advantages of OS-LLMs

- Transparency
- Fine-tuning
- Community

Open-source LLMs not only enable free usage but also impress with transparency by providing insight into architecture, functionality, and training data. Flexibility is enhanced through fine-tuning, allowing customization with proprietary data. Another advantage is the community, fostering continuous improvements and providing a space for the exchange of ideas.

## Use Cases

- Finance
- Mathematics
- Programming
- Argumentation
- Multilingual translation

These language models can be trained for various use cases, with larger models aiming for a balanced mixture of these individual aspects.

> ## 1) BLOOM

Reviewer: No need for enumerating these

```suggestion
## BLOOM
```

> <div style="text-align:center">
>   <img src="https://assets.website-files.com/6139f3cdcbbff3a68486761d/62cce3c835539c54f31329b1_image1.png" alt="Alt text" width="500"/>
> </div>

Reviewer: No need for HTML `<div>` containers to embed images. You can use the Markdown syntax (also see the [assets info box](https://pkeilbach.github.io/htwg-practical-nlp/presentations/presentations/#submission-via-pull-request)

```suggestion
![Bloom Logo](https://assets.website-files.com/6139f3cdcbbff3a68486761d/62cce3c835539c54f31329b1_image1.png)
```

In case you need to adjust the width (but usually the original size works just fine, Mkdocs should adjust the image to the width in case its too big):

```
![Bloom Logo](https://assets.website-files.com/6139f3cdcbbff3a68486761d/62cce3c835539c54f31329b1_image1.png){ width=50% }
```

Please make sure to adjust the ["Alt text"](https://www.w3schools.com/tags/att_img_alt.asp) place holder.

Please adjust that throughout the document.

Our first Large Language Model (LLM) was Bloom, released in July 2022 as a groundbreaking open-source model. Before Bloom, only a few industrial labs could harness the full potential of large language models due to limited resources and rights. Bloom broke this status quo as the first multilingual LLM trained transparently.

With training in 59 languages and an impressive 176 billion parameters, Bloom set a milestone for the accessibility and transparency of large language models.

!!! note
It is noteworthy, however, that this model requires **180 GB of storage** and **360 GB of GPU RAM** for operation.

> <br><br>

Reviewer: No need for these `<br>` tags, can you please remove them throughout the document?

> ---

Reviewer: No need for horizontal lines, could you please remove them throughout the document?

> ## 2) Falcon Series

Reviewer: ```suggestion
## Falcon Series
```

| ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🏆 In September 2023, the TII unveiled the big brother of the Falcon series with **180B** parameters, which, upon its introduction, achieved the highest rankings in most benchmarks. The key to its success lay in its high-quality dataset, which was also released as open source. | ![Alt text](https://gulfbusiness.com/wp-content/uploads/2023/06/UAEs-Falcon-40B-AI-Model-is-Now-Royalty-free-image-credit-Technology-Innovation-Institute.png) |

> #### 2.1 ) The dataset

Reviewer: also watch out for the correct heading level, i.e. `###` here

```suggestion
### The dataset
```

<div style="text-align:center">
  <img src="./assets/os-llms-falcon-dataset.png" alt="Alt text" width="500"/>
</div>

The new dataset **RefinedWeb** represents an innovative, comprehensive web dataset based on CommonCrawl. Falcon enhances this data through deduplication and rigorous content filtering.
In addition to RefinedWeb, books, conversations, code, and technical-scientific papers were also utilized.

The model was primarily trained in **English**, but 7% of the dataset was expanded to **European languages**.

<div style="text-align:center">
  <img src="./assets/os-llms-falcon-refinedweb-eu.png" alt="Alt text" width="500"/>
</div>

> !!! info Required Resources

Reviewer: same as above with the admonition

⚠️ The significance of the parameter count becomes apparent as the training of such models is costly and resource-intensive.

    **7B Model**
    - Training: 384x A100 40GB GPUs for 2 weeks
    - Usage: minimum 16 GB GPU-RAM

    **40B Model**
    - Training: 384x A100 40GB GPUs for 2 months
    - Usage: minimum 90 GB GPU-RAM

    **180B Model**
    - Training: 4,069x A100 40GB GPUs (AWS)

<br><br>

---

> ## 3) LLaMa

Reviewer: ```suggestion
## LLaMa
```

Another success among open-source LLM models is the Llama model by Meta.

<div style="text-align:center">
  <img src="https://mixed.de/wp-content/uploads/2021/11/Meta_Logo.jpg" alt="Alt text" width="200"/>
</div>

> #### 3.1) LLaMA 1

Reviewer: ```suggestion
### LLaMA 1
```

The first Llama model, released on 24.02.2023, achieves performance comparable to OpenAI's GPT-3 model.

Versions:

- **7B**
- **13B**
- **33B**
- **65B**

!!! attention
Unfortunately, the open-source model is limited to non-commercial use.

> #### 3.2) Open LLaMA

Reviewer: ```suggestion
### Open LLaMA
```

<div style="text-align:center">
  <img src="https://sapling.ai/static/img/llms/openllama-logo.png" alt="Alt text" width="150"/>
</div>

In response, a group of students at UC Berkeley in California founded OpenLM Research. Two months later, they released the OpenLLama model based on Meta's LLAMA model.

OpenLLama offers models in versions with **3B**, **7B**, and **13B** parameters. Particularly interesting are the powerful yet "small" V2 models with **3B** and **7B** parameters. These compact yet powerful models were developed from a specially curated dataset.

They are a combination of the Refined Web dataset from Falcon, the Starcoder dataset, and the Redpajama dataset, which is a reproduction of the LLaMA dataset.

<div style="text-align:center">
  <img src="./assets/os-llms-openllama-dataset.png" alt="Alt text" width="600"/>

> </div>

Reviewer: In fact, embedding images using HTML `div` containers does not work (at least on my end):

![image](https://github.com/pkeilbach/htwg-practical-nlp/assets/28002199/90f378c3-76b4-467e-9f4a-fc407443883d)

So I suggest to use this schema:

```md
![Open LLaMA dataset](./assets/os-llms-openllama-dataset.png)
```

> #### 3.2) LLaMA 2

Reviewer: ```suggestion
### LLaMA 2
```

Three months after OpenLLama, Meta releases the LLaMA 2 model, which is open source this time (also for commercial use). Meta's models are impressive, especially the model with 70 billion parameters.

Versions:

- **7B**
- **13B**
- **70B**

> #### LLaMA2 Fine-tuning

Reviewer: ```suggestion
### LLaMA2 Fine-tuning
```
